Feature/better processing

nvdk opened the pull request:
changes the setup of the processing pipeline and limits the amount of concurrent queries to the database. similar to the stop-on-errors branch this pipeline stops when an error occurs.
different from the stop-on-errors branch (#33) we currently only store state at the end of a successful pipeline and ignore members based on their timestamp while processing them. Storing state faster means it's possible we would miss items in the feed when restarting after a db error.  E.g. we store the ldes client state indicating we have fetched a certain page, but haven't actually processed all members of that page.

nvdk committed:
refactor and stop on errors

initially just wanted better state management (storing state faster and stopping
on errors), ended up refactoring a bit. no longer uses the ldes-consumer library
now.

The consumeStream method uses an async promise executor which is considered a
bad pattern, but I think I've handled possible errors and don't really see a
better way of writing this at the moment.

elpoelma committed:
remove ldes-consumer dependency

elpoelma committed:
use nodejs pipeline construct for auto backoff in pipeline

changes the setup of the processing pipeline and limits the amount of concurrent queries to the database
we currently only store state at the end of a succesful pipeline and ignore members based on their timestamp while processing them.

Windvis commented on member-processor.ts:
```diff
@@ -0,0 +1,105 @@
+        }
+      }
+      await timeout(10);
+    }  while (! this.closed);
```
Not 100% sure, but I think this might be an infinite loop since `this.closed` doesn't seem to be set to true anywhere?

nvdk replied:
it should get closed by the pipeline on finishing or on error

cecemel commented:
It doesn't seem to work. On subsequent runs, the consumer stops fetching data from the ldes endpoint.
I briefly tried with `actor-init-ldes-client: 4.0.9`, with some adaptations  (see draft PR: https://github.com/redpencilio/ldes-consumer-service/pull/37), but no success.
The issue is reproducible on multiple endpoints (IPDC and the one in example)

# steps to reproduce
An easy sample config you can start playing with:
```
version: "3.7"
services:
  database:
    image: redpencil/virtuoso:1.0.0
    environment:
      SPARQL_UPDATE: "true"
      DEFAULT_GRAPH: "http://mu.semte.ch/application"
    volumes:
      - ./data/db:/data
    ports:
      - "8890:8890"
  ldes-consumer:
    image: redpencil/ldes-consumer:feature-better-processing
    environment:
      LOG_SPARQL_ALL: "true"
      LDES_STREAM: "https://dev.ldes-endpoint.abb-bfg.s.redpencil.io/felix-debug-1"
      LDES_ENDPOINT_VIEW: "https://dev.ldes-endpoint.abb-bfg.s.redpencil.io/felix-debug-1/1" # first page of the stream
      LDES_RELATION_PATH: "http://www.w3.org/ns/prov#generatedAtTime"
      MU_APPLICATION_GRAPH: "http://mu.semte.ch/graphs/felix-debug-8"
      CRON_PATTERN: "* * * * *" # every minute
      REPLACE_VERSIONS: "false"
```
First time runs fine. Add some data (don't forget to update the URI's)
```
curl -X POST -H "Content-Type: text/turtle" -d '<http://example/10> a <http://xmlns.com/foaf/0.1/Person>.' https://dev.ldes-endpoint.abb-bfg.s.redpencil.io/felix-debug-1?resource=http://example/10
```
Wait for the job to kick in.
No new data will be processed.

nvdk committed:
allow toggling state persistence

we've noticed the client not behaving as we expect when restoring state, so we
disable it for now. can be enabled with a flag when required or for further testing

nvdk commented:
@cecemel I believe the simplest "solution" for now is to not store state: https://github.com/redpencilio/ldes-consumer-service/commit/5d0171a8990856238a73ba383f404730e961beb7

nvdk committed:
Merge branch 'master' into feature/better-processing

cecemel commented:
All right; @nvdk your last commited fixed the reported issue, :+1: 
But; playing a bit further with blank nodes, I notice somehow they keep on being instantiated in a new URI on every run.
Can check it out with sample config from above and this call
```
curl -X POST -H "Content-Type: text/turtle" -d '<http://example/13> a <http://xmlns.com/foaf/0.1/Person>; <http://test> [ a <http://Yes>; <http://bar> "baz"].' https://dev.ldes-endpoint.abb-bfg.s.redpencil.io/felix-debug-1?resource=http://example/13
```
Triples similar to 
```
<http://mu.semte.ch/services/ldes-time-fragmenter/versioned/ec5cd83d-718d-45dc-9f0f-68d5e6266167>	<http://test>	<https://dev.ldes-endpoint.abb-bfg.s.redpencil.io/felix-debug-1/b3730_n3-1> .
```
will keep piling up.

Update v0.7.1:  no problems there (but yeah state is kept, so it's not checked anymore)

cecemel commented:
Would it be a matter of adding an extra filter here, to ignore the already processed ones?
Here https://github.com/redpencilio/ldes-consumer-service/blob/87ec429853c7bbc93dccf4457a27223a1d5c5a88/member-processor.ts#L50 or here
 https://github.com/redpencilio/ldes-consumer-service/blob/87ec429853c7bbc93dccf4457a27223a1d5c5a88/member-processor.ts#L63

cecemel committed:
Change update store logic

When base resource is found and the latest version in the store is
older than the version we are processing, store the newer version.

Note: the logic was present, but only when REPLACE_VERSIONS was set to
true. Hence it was a matter of shuffling variables around

cecemel committed:
Comment and outline.

cecemel committed:
Ensure when fetching latest timestamp, the query sorts results

cecemel committed:
Merge pull request #40 from redpencilio/bugfix/correct-fetching-latest-timestamp

Ensure when fetching latest timestamp, the query sorts results

cecemel committed:
Added warning in case baseResourceUri wasn't found

cecemel committed:
Merge pull request #39 from redpencilio/feature/better-processing-correction-version-handling

Feature/better processing correction version handling
